src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -246,13 +247,66 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
             VisitStatementList(node);
         }
 
+        private struct PossiblyConditionalState
+        {
+            public LocalState State;
+            public LocalState StateWhenTrue;
+            public LocalState StateWhenFalse;
+            public bool IsConditionalState;
+
+            public PossiblyConditionalState(LocalState state)
+            {
+                StateWhenTrue = StateWhenFalse = default;
+                IsConditionalState = false;
+                State = state.Clone();
+            }
+
+            public PossiblyConditionalState(NullableWalker nullableWalker)
+            {
+                if (nullableWalker.IsConditionalState)
+                {
+                    StateWhenTrue = nullableWalker.StateWhenTrue.Clone();
+                    StateWhenFalse = nullableWalker.StateWhenFalse.Clone();
+                    IsConditionalState = true;
+                    State = default;
+                }
+                else
+                {
+                    StateWhenTrue = StateWhenFalse = default;
+                    IsConditionalState = false;
+                    State = nullableWalker.State.Clone();
+                }
+            }
+
+            public PossiblyConditionalState Clone()
+            {
+                PossiblyConditionalState result;
+                if (IsConditionalState)
+                {
+                    result.IsConditionalState = true;
+                    result.State = default;
+                    result.StateWhenTrue = StateWhenTrue.Clone();
+                    result.StateWhenFalse = StateWhenFalse.Clone();
+                }
+                else
+                {
+                    result.IsConditionalState = false;
+                    result.State = State.Clone();
+                    result.StateWhenTrue = default;
+                    result.StateWhenFalse = default;
+                }
+
+                return result;
+            }
+        }
+
         private PooledDictionary<LabelSymbol, (LocalState state, bool believedReachable)>
             LearnFromDecisionDag(
             SyntaxNode node,
             BoundDecisionDag decisionDag,
             BoundExpression expression,
             TypeWithState expressionType,
-            ref LocalState initialState)
+            ref PossiblyConditionalState initialState)
```

Comment:
>PossiblyConditionalState [](start = 16, length = 24)

Is this change necessary? It looks like all callers simply convert a `LocalState` to a `PossiblyConditionalState`. If so, consider converting inside the method.

Reply:
Yes, we need pass in a conditional state as the starting point of DAG analysis. That's how  `IsOk is true` and `IsOk switch { true => ... }` and `switch (IsOk) { ... }` work.

All three callers do `new PossiblyConditionalState(this)` to capture the state after the receiver is evaluated, so we could factor that instance creation here, but I think it's better to let the callers do it, to make it explicit what information is passed to this method.

---
In reply to: [573089929](https://github.com/dotnet/roslyn/pull/51001#discussion_r573089929) [](ancestors = 573089929)